### PR TITLE
Support for specifying filter preferences

### PR DESCRIPTION
**Pull request by danjohnson95**
This PR adds functionality for a user to customise which notifications they receive.

See the full commit message for code explanations.

Users will now see a Settings cog on the top header:
image

This takes the user to the /settings route, which shows currently configured filters and allows for adding:
image

When clicking "Add new filter", the user can specify their criteria
image

image

I've added a quick feature test for this too.

### CODE REVIEW

File: app/Events/JobsPosted.php [hunk 1]
```diff
@@ -2,18 +2,26 @@
...
+    public function __construct(public bool $notifyEmpty = false)
     {
-
+        $this->jobs = JobPost::visible()
+            ->unnotified()
+            ->filtered()
+            ->orderBy('published_at', 'desc')
+            ->get();
     }
 }
```
Comment (LukeTowers) [inline app/Events/JobsPosted.php]:
```suggestion
}

```

File: app/Http/Livewire/NewFilter.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        ]);
+
+        $this->emit('filterAdded');
+    }
+
+    public function render()
+    {
+        return view('livewire.new-filter');
+    }
+}
```
Comment (LukeTowers) [inline app/Http/Livewire/NewFilter.php]:
```suggestion
}

```

File: app/Http/Livewire/Settings.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        $this->closeFilterWindow();
+        $this->filters = Filter::orderBy('created_at')->get();
+    }
+
+    public function closeFilterWindow()
+    {
+        $this->addNewFilter = false;
+    }
+}
```
Comment (LukeTowers) [inline app/Http/Livewire/Settings.php]:
```suggestion
}

```

File: app/Jobs/FetchNewJobs.php [hunk 4]
```diff
@@ -61,9 +61,20 @@ public function handle(): void
...
+        $this->notifyJobsPosted();
+    }
+
+    /**
+     * Notify the user of new jobs that have been posted.
+     * If the user has defined filters for their notification preference,
+     * we'll only notify them of jobs that match their criteria.
+     *
+     * The event constructor is responsible for fetching the
+     * jobs which meet the criteria.
```
Comment (LukeTowers) [inline app/Jobs/FetchNewJobs.php]: Why is the event constructor responsible for gathering the data? It seems backwards to me, you wouldn't be able to reuse the event for anything else taking that approach.

What are the benefits of making the event prepare its own data instead of passing the relevant data to the event when creating it? I'm not sure I've seen that pattern before so I'm genuinely curious.

File: app/Listeners/SendNewJobsNotification.php [hunk 5]
```diff
@@ -4,56 +4,78 @@
...
+     * This is only called if the notifyEmpty flag is set to true.
+     */
+    protected function handleEmptyNotification(): void
+    {
+        Notification::new()
+            ->title("No new jobs")
+            ->message("There are no new jobs available.")
+            ->event(self::EVENT_CLICK_PREFIX . 'nojobs')
+            ->show();
+    }
```
Comment (LukeTowers) [inline app/Listeners/SendNewJobsNotification.php:80]: Good refactoring job on this class, I like it ❤️

File: app/Models/Filter.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                return 'doesn\'t contain';
+            default:
+                return $this->operation;
+        }
+    }
+
+    /**
+     * Transforms the filter to a query arguments
+    */
+    public function operationToQuery()
```
Comment (LukeTowers) [inline app/Models/Filter.php]:
```suggestion
    public function operationToQuery(): array
```

File: resources/views/components/filter.blade.php [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<div>
+    <span class="font-medium text-primary">{{ $filter->getFriendlyField() }}</span>
+    <span class="text-gray-500">{{ ' ' . $filter->getFriendlyOperation() . ' ' }}</span>
+    <code class="font-mono">{{ $filter->query }}</code>
+</div>
```
Comment (LukeTowers) [inline resources/views/components/filter.blade.php]:
```suggestion
</div>

```

File: resources/views/components/footer.blade.php [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+        <div class="px-5 py-2">
+            <a onclick="shell.openExternal('https://twitter.com/laraveljobs')" href="javascript:;" class="text-gray-500 hover:text-gray-900"> Twitter </a>
+        </div>
+    </nav>
+    <p class="mt-4 text-center text-gray-400">
+        Built by <a onclick="shell.openExternal('https://userscape.com')" href="javascript:;">UserScape</a> in partnership with <a onclick="shell.openExternal('https://laravel-news.com')" href="javascript:;">Laravel News</a>
+    </p>
+    <p class="mt-1 text-center text-gray-400">&copy; 2014 - {{ date('Y') }}</p>
+</footer>
```
Comment (LukeTowers) [inline resources/views/components/footer.blade.php]:
```suggestion
</footer>

```

File: resources/views/components/header.blade.php [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+                <path stroke-linecap="round" stroke-linejoin="round" d="M15 12a3 3 0 11-6 0 3 3 0 016 0z" />
+            </svg>
+        </a>
+    @endif
+
+    {{-- <div class="">
+        <a href="https://larajobs.com/laravel-consultants" class="px-4 py-2 rounded-sm border-2 border-accent bg-sky-200 text-zinc-800 hover:bg-white hover:text-accent">Hire a Consultant</a>
+        <a href="https://larajobs.com/create" class="px-4 py-2 rounded-sm border-2 border-accent bg-accent text-white hover:bg-white hover:text-accent">Post a Job</a>
+    </div> --}}
+</header>
```
Comment (LukeTowers) [inline resources/views/components/header.blade.php]:
```suggestion
</header>

```

File: resources/views/components/layout.blade.php [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+            <main>
+                {{ $slot }}
+            </main>
+
+            <x-footer/>
+        </div>
+        @livewireScripts
+    </body>
+</html>
```
Comment (LukeTowers) [inline resources/views/components/layout.blade.php]:
```suggestion
</html>

```

File: resources/views/livewire/new-filter.blade.php [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+                @error('field') <span class="text-sm text-red-500">{{ $message }}</span> @enderror
+            </div>
+            <div>
+                <label class="text-sm text-gray-500 mb-2">Operation</label>
+                <select wire:model="operation" class="border border-gray-300 rounded-md w-full p-2">
+                    <option>-- Select an operation --</option>
+                    <option value="{{ App\Enums\FilterOperation::Equals }}">equals</option>
+                    <option value="{{ App\Enums\FilterOperation::NotEquals }}">is not</option>
+                    <option value="{{ App\Enums\FilterOperation::Contains }}">contains</option>
+                    <option value="{{ App\Enums\FilterOperation::NotContains }}">doesnt contain</option>
```
Comment (LukeTowers) [inline resources/views/livewire/new-filter.blade.php]:
```suggestion
                    <option value="{{ App\Enums\FilterOperation::NotContains }}">doesn't contain</option>
```

File: resources/views/livewire/new-filter.blade.php [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+                @error('query') <span class="text-sm text-red-500">{{ $message }}</span> @enderror
+            </div>
+            <div>
+                <button type="submit" class="bg-primary text-white font-medium rounded-lg w-full p-3">
+                    Add filter
+                </button>
+            </div>
+        </div>
+    </div>
+</div>
```
Comment (LukeTowers) [inline resources/views/livewire/new-filter.blade.php]:
```suggestion
</div>

```

File: resources/views/livewire/settings.blade.php [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+                <path d="M10.75 4.75a.75.75 0 00-1.5 0v4.5h-4.5a.75.75 0 000 1.5h4.5v4.5a.75.75 0 001.5 0v-4.5h4.5a.75.75 0 000-1.5h-4.5v-4.5z" />
+            </svg>
+            <span>Add {{$filters->isNotEmpty() ? 'another ' : '' }}filter</span>
+        </button>
+    </section>
+
+    @if ($addNewFilter)
+        <livewire:new-filter/>
+    @endif
+</div>
```
Comment (LukeTowers) [inline resources/views/livewire/settings.blade.php]:
```suggestion
</div>

```

File: resources/views/menubar/index.blade.php [hunk 14]
```diff
@@ -1,121 +1,64 @@
...
-                    <p class="mt-1 text-center text-gray-400">&copy; 2014 - {{ date('Y') }}</p>
+                    </div>
                 </div>
             </div>
-        </div>
-    </body>
-</html>
+        </a>
+    @endforeach
+</x-layout>
```
Comment (LukeTowers) [inline resources/views/menubar/index.blade.php]:
```suggestion
</x-layout>

```

File: routes/web.php [hunk 15]
```diff
@@ -19,3 +20,5 @@
 
     return view('menubar.index', ['jobPosts' => $jobs]);
 })->name('menubar.index');
+
+Route::get('/settings', Settings::class)->name('menubar.settings');
```
Comment (LukeTowers) [inline routes/web.php]:
```suggestion
Route::get('/settings', Settings::class)->name('menubar.settings');

```

File: tests/Feature/JobsPostedEventTest.php [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+        $event = new JobsPosted();
+
+        // Unfortunately we can't mock the NativePHP Notification
+        // facade, so instead we're just going to ensure that the
+        // jobs within the event contain what we're expecting to
+        // notify the user about.
+        $this->assertNotContains($job->id, $event->jobs->pluck('id'));
+    }
+}
```
Comment (LukeTowers) [inline tests/Feature/JobsPostedEventTest.php]:
```suggestion
}

```

**Review by LukeTowers — CHANGES_REQUESTED** [review]
Looks great @danjohnson95! Would you be willing to add a GitHub action to run the test suite as well? If not I should be able to take care of that

File: app/Jobs/FetchNewJobs.php (hunk 4 above)
Comment (danjohnson95) [reply app/Jobs/FetchNewJobs.php]: It's a good observation - I mainly did this to increase testability. I wanted our test to be able to prove that only jobs meeting the filter criteria are notified.

If we're passing the jobs to the event, we can't reliably test _which_ jobs the user will be notified of when `FetchNewJobs` is run, unless we were to copy over the logic of the query into the test (but then we'd end up with a brittle test)

I think we've got three options:

- Make the event responsible for fetching the jobs. (The change made in this PR). This way, both our tests and the real implementation will be using the same logic to fetch jobs to notify the user about. As you've mentioned, the downside of this is that the event is only usable for notifying the user of new jobs that meet their criteria and haven't already been notified of yet.
- Make the listener responsible for fetching the jobs. This would give us the same result as the above option, but at least we could re-use the event by attaching different listeners.
- Refactor the FetchNewJobs job so that it has a service responsible for fetching the data from the RSS feed. That way, we can mock the response of the RSS feed in our tests. I think that's a good change to make anyway because it'll allow us to test the persistence part of this job, but it still doesn't really give us a way to test the notification sending part _in isolation_.

What do you think? I'm leaning towards option three. We'll make the `JobsPostedEventTest` responsible for testing that notifications are sent (not _which_ jobs the user is notified of), and we'll make a new test for `FetchNewJobs` which is responsible for testing that the jobs are persisted correctly _and_ the event is dispatched with only jobs the user is interested in.

**TLDR**

- did it for testability
- thought the event should be responsible for determining if the user wants the notification.
- now I've changed my mind, I'll refactor and make a new test for testing that bit

**Comment by danjohnson95** [comment]
Thanks for the feedback! I've added a workflow file to run the tests but it looks like it's not triggering - maybe the repo needs GitHub Actions enabled in the settings on your end? Let me know if not - it's also possible I've fudged the workflow file haha

File: app/Jobs/FetchNewJobs.php (hunk 4 above)
Comment (LukeTowers) [reply app/Jobs/FetchNewJobs.php]: The third option sounds great, thanks for thinking on it further and coming up with that as the solution :)

File: app/Jobs/FetchNewJobs.php [hunk 17]
```diff
@@ -61,8 +63,22 @@ public function handle(): void
         }
 
         // Fire the event
-        // @TODO: Apply filters here
-        $jobs = JobPost::visible()->unnotified()->orderBy('published_at', 'desc')->get()->all();
+        $this->notifyJobsPosted();
```
Comment (LukeTowers) [inline app/Jobs/FetchNewJobs.php:62]: If we're not doing anything with the $posts array then we should probably get rid of it now

File: tests/Feature/FetchNewJobsTest.php [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+
+        // Dispatch the job
+        FetchNewJobs::dispatch();
+
+        // Assert that the event was dispatched with no jobs
+        Event::assertDispatched(JobsPosted::class, function ($event) {
+            return $event->jobs->count() === 0;
+        });
+    }
+}
```
Comment (LukeTowers) [inline tests/Feature/FetchNewJobsTest.php]:
```suggestion
}

```

**Review by LukeTowers — APPROVED** [review]

File: app/Jobs/FetchNewJobs.php (hunk 17 above)
Comment (danjohnson95) [reply app/Jobs/FetchNewJobs.php:62]: Good spot, thanks. It's now removed 👍
